crates/bevy_time/src/time.rs

```diff
@@ -1,218 +1,288 @@
+/// and passing it to [`new_with()`](Time::new_with). These clocks can be
+/// inserted as resources as normal and then accessed by systems. You can use
+/// the [`advance_by()`](Time::advance_by) or [`advance_to()`](Time::advance_to)
+/// methods to move the clock forwards based on your own logic.
+///
+/// If you want to add methods for your time instance and they require access to
+/// both your context and the generic time part, it's probably simplest to add a
+/// custom trait for them and implement it for `Time<Custom>`.
+///
+/// Your context struct will need to implement the
+/// [`Default`](std::default::Default) trait because [`Time`] structures support
+/// reflection. It also makes initialization trivial by being able to call
+/// `app.init_resource::<Time<Custom>>()`.
+///
+/// You can also replace the "generic" `Time` clock resource if the "default"
+/// time for your game should not be the default virtual time provided. You can
+/// get a "generic" snapshot of your clock by calling `as_generic()` and then
+/// overwrite the [`Time`] resource with it. The default systems added by
+/// [`TimePlugin`](crate::TimePlugin) will overwrite the [`Time`] clock during
+/// [`First`](bevy_app::First) and [`FixedUpdate`](bevy_app::FixedUpdate)
+/// schedules.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # use bevy_time::prelude::*;
+/// # use bevy_utils::Instant;
+/// #
+/// #[derive(Debug)]
+/// struct Custom {
+///     last_external_time: Instant,
+/// }
+///
+/// impl Default for Custom {
+///     fn default() -> Self {
+///         Self {
+///             last_external_time: Instant::now(),
+///         }
+///     }
+/// }
+///
+/// trait CustomTime {
+///     fn update_from_external(&mut self, instant: Instant);
+/// }
+///
+/// impl CustomTime for Time<Custom> {
+///     fn update_from_external(&mut self, instant: Instant) {
+///          let delta = instant - self.context().last_external_time;
+///          self.advance_by(delta);
+///          self.context_mut().last_external_time = instant;
+///     }
+/// }
+/// ```
+#[derive(Resource, Debug, Copy, Clone, Reflect)]
 #[reflect(Resource, Default)]
-pub struct Time {
-    startup: Instant,
-    first_update: Option<Instant>,
-    last_update: Option<Instant>,
-    // pausing
-    paused: bool,
-    // scaling
-    relative_speed: f64, // using `f64` instead of `f32` to minimize drift from rounding errors
+pub struct Time<T: Default = ()> {
+    context: T,
+    wrap_period: Duration,
     delta: Duration,
     delta_seconds: f32,
     delta_seconds_f64: f64,
     elapsed: Duration,
     elapsed_seconds: f32,
     elapsed_seconds_f64: f64,
-    raw_delta: Duration,
-    raw_delta_seconds: f32,
-    raw_delta_seconds_f64: f64,
-    raw_elapsed: Duration,
-    raw_elapsed_seconds: f32,
-    raw_elapsed_seconds_f64: f64,
-    // wrapping
-    wrap_period: Duration,
     elapsed_wrapped: Duration,
     elapsed_seconds_wrapped: f32,
     elapsed_seconds_wrapped_f64: f64,
```

Comment:
Since this cache is recalculated whenever time advances, no performance improvement can be expected. It is okay to retain only the pre-calculated value.

Reply:
Comments in pull discussion on this as well. I'd be happy to remove precalculated cache, and even happer would be to get rid of the overly duplicated API, but I didn't want to make that decision in this pull, since it's also controversial.

Reply:
The difference is so very small that it cannot be measured in the benchmark, so no one will notice if you change this silently. 🏃 

The current code computes `6 * 3` times per frame, which may be slower than not caching.

Reply:
There's four choices:
1. Leave as is, figure out something in another pull if it feels meaningful.
2. Remove precomputation, but keep the same API, so `delta_seconds()` will return `f32`, just calculated on every request.
3. Create `CachedDuration` which keeps precomputation, and change API to return only `CachedDuration`, forcing every user to change `time.delta_seconds()` to `time.delta().as_secs_f32()` but keeping same performance "advantage".
4. Remove precomputation and change API to return only `Duration`. Every user needs to switch from `time.delta_seconds()` to `time.delta().as_secs_f32()`, but it will be the Rust standard `Duration` class so it should be quite obvious. (Not sure what to do about wrapped time here...)

In a way, I do understand the precomputation – even if there's `6 * 3` computation per frame to get the values, most of which will not be used, it's still a meaningless amount of time. Where as if there is one `time.delta_seconds()` call in a system which does collision detection, for example, then that's easily millions of conversions to float per frame... if the compiler for some reason doesn't optimize it outside the loops.

But I think what you are proposing is number two? I'm fine with switching this pull to do that, if I can count on not having to revert my changes in a few days :-)